src/lsm/manifest.zig

```diff
@@ -521,25 +525,115 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
             manifest: *const Manifest,
             key_min: Key,
             key_max: Key,
+            options: struct { value_count: u32 },
         ) CompactionRange {
             assert(key_min <= key_max);
+            assert(options.value_count > 0);
+            assert(options.value_count <= Table.value_count_max);
+
             const level_b = 0;
             const manifest_level: *const Level = &manifest.levels[level_b];
+            assert(manifest_level.table_count_visible <= growth_factor);
 
             // We are guaranteed to get a non-null range because Level 0 has
             // lsm_growth_factor number of tables, so the number of tables that intersect
             // with the immutable table can be no more than lsm_growth_factor.
-            const range = manifest_level.tables_overlapping_with_key_range(
+            const range_overlap = manifest_level.tables_overlapping_with_key_range(
                 key_min,
                 key_max,
                 snapshot_latest,
                 growth_factor,
             ).?;
 
+            // Attempt to coalesce with adjacent tables in level 0.
+            const range_coalesced = range: {
+                const value_count_target = stdx.div_ceil((Table.value_count_max *
+                    constants.lsm_table_coalescing_threshold_percent), 100);
+                assert(value_count_target > 1);
+                assert(value_count_target < Table.value_count_max);
+
+                var value_count_output: u32 = options.value_count;
+                for (range_overlap.tables.const_slice()) |*table| {
+                    value_count_output += table.table_info.value_count;
+                }
+
+                // Set to true when we encounter a coalesce-able table that is small enough to
+                // warrant coalescing.
+                var coalesced_small_table: bool = value_count_output < value_count_target;
+
+                var range = range_overlap;
+                outer: for ([_]Direction{ .descending, .ascending }) |direction| {
+                    inner: for (0..constants.lsm_growth_factor + 1) |_| {
```

Comment:
I'm not sure I understand the bounds here correctly - assuming `range` has a single table in it, the most iterations we would need to do here would be `manifest_level.table_count_visible - 1` (or `constants.lsm_growth_factor - 1`)?

Reply:
In the case of coalescing, `range` may initially overlap with no tables at all. But we are still willing to coalesce with up to `growth_factor` tables.

Reply:
That makes sense! And removes the `- 1`, but wouldn't the bounds then be `0..constants.lsm_growth_factor`? :thinking: